rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -180,27 +171,25 @@ class ReplayRSpace[F[_]: Concurrent: ContextShift: Log: Metrics: Span, C, P, A,
       c =>
         store
           .getData(c)
-          .map(_.iterator.zipWithIndex.toSeq)
-          .map(
-            as => {
-              c -> {
-                if (c == channel)
-                  (Datum(data, persist, produceRef), -1) +: as
-                else as
-              }.filter {
-                matches(comm)
-              }
-            }
-          )
+          // Convert to immutable.Seq because of defined Traversable instance
+          .map(_.iterator.zipWithIndex.to[immutable.Seq])
+          .flatMap { datums =>
+            (if (c == channel) (Datum(data, persist, produceRef), -1) +: datums else datums)
+              .filterA(matches(comm))
+              .map(_.to[Seq])
+          }
+          .map((c, _))
     )
 
-  private[this] def matches(comm: COMM)(datumWithIndex: (Datum[A], _)): Boolean = {
+  private[this] def matches(comm: COMM)(datumWithIndex: (Datum[A], _)): F[Boolean] = {
     val datum: Datum[A] = datumWithIndex._1
-    def wasRepeatedEnoughTimes: Boolean =
-      if (!datum.persist) {
-        comm.timesRepeated(datum.source) === produceCounter.get(datum.source)
-      } else true
-    comm.produces.contains(datum.source) && wasRepeatedEnoughTimes
+    def wasRepeatedEnoughTimes: F[Boolean] =
+      produceCounter.get.map { counter =>
+        if (!datum.persist) {
+          comm.timesRepeated(datum.source) === counter(datum.source)
+        } else true
+      }
+    Sync[F].delay(comm.produces.contains(datum.source)) &&^ wasRepeatedEnoughTimes
```

Comment:
The subtle issue here is that second expression in boolean && operation must be evaluated lazily.